pkg/alertmanager/multitenant.go

```diff
@@ -944,6 +1083,54 @@ func (am *MultitenantAlertmanager) UpdateState(ctx context.Context, part *cluste
 	return &alertmanagerpb.UpdateStateResponse{Status: alertmanagerpb.OK}, nil
 }
 
+// deleteUnusedLocalUserState deletes local files for users that we no longer need.
+func (am *MultitenantAlertmanager) deleteUnusedLocalUserState() {
+	userDirs := am.getPerUserDirectories()
+
+	// And delete remaining files.
+	for userID, dir := range userDirs {
+		am.alertmanagersMtx.Lock()
+		userAM := am.alertmanagers[userID]
+		am.alertmanagersMtx.Unlock()
+
+		// Don't delete directory if AM for user still exists.
+		if userAM != nil {
+			continue
+		}
+
+		err := os.RemoveAll(dir)
+		if err != nil {
+			level.Warn(am.logger).Log("msg", "failed to delete directory for user", "dir", dir, "user", userID, "err", err)
+		} else {
+			level.Info(am.logger).Log("msg", "deleted local directory for user", "dir", dir, "user", userID)
+		}
+	}
+}
+
+// getPerUserDirectories returns map of users to their directories (full path). Only users with local
+// directory are returned.
+func (am *MultitenantAlertmanager) getPerUserDirectories() map[string]string {
+	files, err := ioutil.ReadDir(am.cfg.DataDir)
+	if err != nil {
+		level.Warn(am.logger).Log("msg", "failed to list local dir", "dir", am.cfg.DataDir, "err", err)
+		return nil
+	}
+
+	result := map[string]string{}
+
+	for _, f := range files {
+		fullPath := filepath.Join(am.cfg.DataDir, f.Name())
+
+		if !f.IsDir() {
+			level.Warn(am.logger).Log("msg", "ignoring local file", "file", fullPath)
```

Comment:
I would be more descriptive, otherwise it's difficult to understand what this log message is about:

```suggestion
			level.Warn(am.logger).Log("msg", "ignoring unexpected file while scanning local alertmanager configs", "file", fullPath)
```